File: apps/visualizer/backend/api/schemas/jobs.py
"""Jobs API and socket payload models — single source of truth for Job shapes."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, RootModel

JobStatus = Literal['pending', 'running', 'completed', 'failed', 'cancelled']
JobLogLevel = Literal['debug', 'info', 'warning', 'error']
ErrorSeverity = Literal['warning', 'error', 'critical']
LibraryDbSource = Literal['env', 'config', 'default', 'none']

_CHECKPOINT_LIST_KEYS = (
    'processed_pairs',
    'processed_media_keys',
    'processed_image_keys',
    'processed_triplets',
)


class JobLog(BaseModel):
    timestamp: str
    level: JobLogLevel
    message: str


class Job(BaseModel):
    """Shared by REST job endpoints and ``job_updated`` / ``job_created`` socket emits."""

    model_config = ConfigDict(extra='forbid')

    id: str
    type: str
    status: JobStatus
    progress: int
    current_step: str | None = None
    logs: list[JobLog] = Field(default_factory=list)
    logs_total: int = 0
    warning_count: int = 0
    error_count: int = 0
    last_log_at: str | None = None
    result: Any | None = None
    error: str | None = None
    error_severity: ErrorSeverity | None = None
    created_at: str
    started_at: str | None = None
    completed_at: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    cancel_noop: bool | None = None
    cancel_noop_reason: str | None = None


class JobCreateRequest(BaseModel):
    type: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class PaginationMeta(BaseModel):
    offset: int
    limit: int
    current_page: int
    total_pages: int
    has_more: bool


class JobsListResponse(BaseModel):
    total: int
    data: list[Job]
    pagination: PaginationMeta


class JobListResponse(RootModel[list[Job]]):
    """``GET /api/jobs/active`` response body."""


class LibraryDbInfo(BaseModel):
    path: str | None
    source: LibraryDbSource
    exists: bool
    reason: str | None = None


class JobsHealth(BaseModel):
    library_db: LibraryDbInfo
    jobs_requiring_catalog: list[str]
    catalog_available: bool


class JobsProcessorHealth(BaseModel):
    running: bool
    started_at: float | None = None
    last_iteration_at: float | None = None
    last_iteration_age_seconds: float | None = None
    iterations_total: int = 0
    current_job_id: str | None = None
    current_job_started_at: float | None = None
    pending_count: int = 0
    running_count: int = 0
    stale: bool = False
    stale_threshold_seconds: float
    last_error: str | None = None


class JobsRecoveredPayload(BaseModel):
    job_ids: list[str]


class ErrorBody(BaseModel):
    error: str
    code: str | None = None


class CatalogUnavailableError(ErrorBody):
    code: Literal['catalog_unavailable']
    library_db: LibraryDbInfo


class DbBusyError(ErrorBody):
    code: Literal['db_busy']


def compact_checkpoint_lists(checkpoint: dict) -> dict:
    compact = dict(checkpoint)
    for key in _CHECKPOINT_LIST_KEYS:
        value = compact.get(key)
        if isinstance(value, list):
            compact[f'{key}_count'] = len(value)
            del compact[key]
    return compact


def compact_job_payload(job: dict) -> dict:
    """Replace bulky checkpoint lists with ``*_count`` tallies for wire payloads."""
    payload = dict(job)
    metadata = payload.get('metadata')
    if isinstance(metadata, dict):
        checkpoint = metadata.get('checkpoint')
        if isinstance(checkpoint, dict):
            payload['metadata'] = {
                **metadata,
                'checkpoint': compact_checkpoint_lists(checkpoint),
            }
    return payload


def enrich_job_log_stats(db, job: dict) -> dict:
    """Attach log summary fields from ``job_logs`` (same as list/active endpoints)."""
    from database import get_job_log_stats_bulk

    enriched = dict(job)
    stats = get_job_log_stats_bulk(db, [enriched['id']]).get(enriched['id'], {})
    enriched['logs_total'] = int(stats.get('logs_total', enriched.get('logs_total', 0) or 0))
    enriched['warning_count'] = int(stats.get('warning_count', enriched.get('warning_count', 0) or 0))
    enriched['error_count'] = int(stats.get('error_count', enriched.get('error_count', 0) or 0))
    enriched['last_log_at'] = stats.get('last_log_at', enriched.get('last_log_at'))
    return enriched


def validate_job_payload(job: dict) -> dict:
    """Validate a compacted job dict against :class:`Job`; raises on shape drift."""
    data = Job.model_validate(compact_job_payload(job)).model_dump(mode='json')
    if data.get('cancel_noop') is None:
        data.pop('cancel_noop', None)
        data.pop('cancel_noop_reason', None)
    return data


def build_job_emit_payload(db, job: dict, **extra) -> dict:
    """Prepare and validate a job dict for socket emit or JSON response."""
    payload = enrich_job_log_stats(db, job)
    payload.update(extra)
    return validate_job_payload(payload)


def validate_jobs_recovered_payload(payload: dict) -> dict:
    """Validate ``jobs_recovered`` socket payload."""
    return JobsRecoveredPayload.model_validate(payload).model_dump(mode='json')

File: apps/visualizer/backend/src/api/schemas/jobs.ts
/**
 * Jobs API and socket payload models.
 *
 * `Job` is shared by the REST endpoints and the `job_created` / `job_updated`
 * socket emits, which is the point: the frontend applies a socket payload straight
 * into the same state a REST response populated, so the two must be the same shape.
 */
import { z } from '@hono/zod-openapi';
import { PaginationMeta } from './pagination.js';

export const JobStatus = z.enum(['pending', 'running', 'completed', 'failed', 'cancelled']);
export const JobLogLevel = z.enum(['debug', 'info', 'warning', 'error']);
export const ErrorSeverity = z.enum(['warning', 'error', 'critical']);
export const LibraryDbSource = z.enum(['env', 'config', 'default', 'none']);

/** Not `.strict()`, matching the Python model, which sets no `extra` policy. */
export const JobLog = z
  .object({
    timestamp: z.string(),
    level: JobLogLevel,
    message: z.string(),
  })
  .openapi('JobLog');

export const Job = z
  .object({
    id: z.string(),
    type: z.string(),
    status: JobStatus,
    progress: z.int(),
    current_step: z.string().nullish(),
    logs: z.array(JobLog).default([]),
    logs_total: z.int().default(0),
    warning_count: z.int().default(0),
    error_count: z.int().default(0),
    last_log_at: z.string().nullish(),
    result: z.unknown().nullish(),
    error: z.string().nullish(),
    error_severity: ErrorSeverity.nullish(),
    created_at: z.string(),
    started_at: z.string().nullish(),
    completed_at: z.string().nullish(),
    metadata: z.record(z.string(), z.unknown()).default({}),
    /** Present only on a cancel that changed nothing. */
    cancel_noop: z.boolean().nullish(),
    cancel_noop_reason: z.string().nullish(),
  })
  .strict()
  .openapi('Job');

export const JobCreateRequest = z
  .object({
    type: z.string(),
    metadata: z.record(z.string(), z.unknown()).default({}),
  })
  .openapi('JobCreateRequest');

export const JobsListResponse = z
  .object({
    total: z.int(),
    data: z.array(Job),
    pagination: PaginationMeta,
  })
  .openapi('JobsListResponse');

/** `GET /api/jobs/active` returns a bare array. */
export const JobListResponse = z.array(Job).openapi('JobListResponse');

export const LibraryDbInfo = z
  .object({
    path: z.string().nullable(),
    source: LibraryDbSource,
    exists: z.boolean(),
    reason: z.string().nullish(),
  })
  .openapi('LibraryDbInfo');

export const JobsHealth = z
  .object({
    library_db: LibraryDbInfo,
    jobs_requiring_catalog: z.array(z.string()),
    catalog_available: z.boolean(),
  })
  .openapi('JobsHealth');

export const JobsProcessorHealth = z
  .object({
    running: z.boolean(),
    started_at: z.number().nullish(),
    last_iteration_at: z.number().nullish(),
    last_iteration_age_seconds: z.number().nullish(),
    iterations_total: z.int().default(0),
    current_job_id: z.string().nullish(),
    current_job_started_at: z.number().nullish(),
    pending_count: z.int().default(0),
    running_count: z.int().default(0),
    stale: z.boolean().default(false),
    stale_threshold_seconds: z.number(),
    last_error: z.string().nullish(),
  })
  .openapi('JobsProcessorHealth');

export const JobsRecoveredPayload = z
  .object({ job_ids: z.array(z.string()) })
  .openapi('JobsRecoveredPayload');

/**
 * `ErrorBody` subclasses, spelled out rather than built with `.extend()`.
 *
 * pydantic inlines an inherited field into the subclass schema, so Flask emits a
 * flat object. Extending the *registered* `ErrorBody` here would emit
 * `allOf: [$ref ErrorBody, { code }]` instead, and the contract diff would see a
 * different shape. The `error` field is repeated for that reason, not by oversight.
 */
export const CatalogUnavailableError = z
  .object({
    error: z.string(),
    code: z.literal('catalog_unavailable'),
    library_db: LibraryDbInfo,
  })
  .openapi('CatalogUnavailableError');

export const DbBusyError = z
  .object({
    error: z.string(),
    code: z.literal('db_busy'),
  })
  .openapi('DbBusyError');

export type Job = z.infer<typeof Job>;

/**
 * Keys whose checkpoint lists are replaced by a `*_count` tally on the wire.
 *
 * A resumed batch checkpoint holds every processed key, which for a 43,000-image
 * embed job is a multi-megabyte array. The UI only ever shows how many.
 */
const CHECKPOINT_LIST_KEYS = [
  'processed_pairs',
  'processed_media_keys',
  'processed_image_keys',
  'processed_triplets',
] as const;

export function compactCheckpointLists(
  checkpoint: Record<string, unknown>,
): Record<string, unknown> {
  const compact = { ...checkpoint };
  for (const key of CHECKPOINT_LIST_KEYS) {
    const value = compact[key];
    if (Array.isArray(value)) {
      compact[`${key}_count`] = value.length;
      delete compact[key];
    }
  }
  return compact;
}

/** Replace bulky checkpoint lists with tallies, for a wire payload. */
export function compactJobPayload<T extends Record<string, unknown>>(job: T): T {
  const metadata = job.metadata;
  if (!metadata || typeof metadata !== 'object' || Array.isArray(metadata)) return job;
  const checkpoint = (metadata as Record<string, unknown>).checkpoint;
  if (!checkpoint || typeof checkpoint !== 'object' || Array.isArray(checkpoint)) return job;
  return {
    ...job,
    metadata: {
      ...(metadata as Record<string, unknown>),
      checkpoint: compactCheckpointLists(checkpoint as Record<string, unknown>),
    },
  };
}
